src/java.base/share/classes/jdk/internal/foreign/SegmentFactories.java

```diff
@@ -175,39 +176,67 @@ public static OfDouble arrayOfDoubleSegment(Object base, long offset, long lengt
         return new OfDouble(offset, base, length, readOnly, bufferScope);
     }
 
-    public static NativeMemorySegmentImpl allocateSegment(long byteSize, long byteAlignment, MemorySessionImpl sessionImpl,
-                                                          boolean shouldReserve) {
+    public static long allocateNative(long byteSize, long byteAlignment, MemorySessionImpl sessionImpl,
+                                      boolean shouldReserve, boolean init) {
         ensureInitialized();
         sessionImpl.checkValidState();
         if (VM.isDirectMemoryPageAligned()) {
             byteAlignment = Math.max(byteAlignment, AbstractMemorySegmentImpl.NIO_ACCESS.pageSize());
         }
-        long alignedSize = Math.max(1L, byteAlignment > MAX_MALLOC_ALIGN ?
-                byteSize + (byteAlignment - 1) :
-                byteSize);
+        // Align up to 8 so we can init the memory with longs
+        long alignedSize = init ? Utils.alignUp(byteSize, Long.BYTES) : byteSize;
+
+        if (byteAlignment > MAX_MALLOC_ALIGN) {
+            return allocateNativeOveraligned(byteSize, byteAlignment, sessionImpl, shouldReserve, init, alignedSize);
+        }
 
         if (shouldReserve) {
             AbstractMemorySegmentImpl.NIO_ACCESS.reserveMemory(alignedSize, byteSize);
         }
-
-        long buf = allocateMemoryWrapper(alignedSize);
-        long alignedBuf = Utils.alignUp(buf, byteAlignment);
-        NativeMemorySegmentImpl segment = new NativeMemorySegmentImpl(buf, alignedSize,
-                false, sessionImpl);
+        long address = allocateMemoryWrapper(alignedSize);
+        if (init) {
+            initNativeMemory(address, alignedSize);
+        }
         sessionImpl.addOrCleanupIfFail(new MemorySessionImpl.ResourceList.ResourceCleanup() {
             @Override
             public void cleanup() {
-                UNSAFE.freeMemory(buf);
+                UNSAFE.freeMemory(address);
                 if (shouldReserve) {
                     AbstractMemorySegmentImpl.NIO_ACCESS.unreserveMemory(alignedSize, byteSize);
                 }
             }
         });
-        if (alignedSize != byteSize) {
-            long delta = alignedBuf - buf;
-            segment = (NativeMemorySegmentImpl) segment.asSlice(delta, byteSize);
+        return address;
+    }
+
+    @DontInline
+    private static long allocateNativeOveraligned(long byteSize, long byteAlignment, MemorySessionImpl sessionImpl,
```

Comment:
This would always make a the session escape in the case of over aligned allocation. Maybe it's possible to move the call to `addCleanupIfFail` to `allocateNative` instead? (that seems to be the only use of `sessionImpl` here).

Reply:
In fact, except for the `alignUp` call, I'm not sure I see enough difference between the two code paths to warrant splitting out the code to a new method? AFAIU, the split was made to avoid playing with slicing - but if we create the segment as the very last thing we do in `allocateNative`, do we care?

Reply:
That's fair, I have moved it back to `SegmentFactories::allocateNative` and get rid of this method.